Fix #6819: add check image size <10MB

matthyx opened the pull request:
Try to get "Content-Length" header for image and make sure it's smaller than 10MB.
Rely on loop in handle to retry if we fail... btw why is that 3 times for cats vs 5 times for dogs?

BenTheElder commented:
> Rely on loop in handle to retry if we fail... btw why is that 3 times for cats vs 5 times for dogs?

different people wrote them, I put in 5 for dogs because the API is quick and we sometimes have to discard "working" results due to being a file type github doesn't support.

BenTheElder commented:
/ok-to-test

matthyx commented:
@cjwagner I have moved the code into github/helpers.go
@BenTheElder I have updated dog_test.go is it OK to use real images from random.dog? If that's fine, I will refactor cat_test.go the same way...

cjwagner commented on prow/github/helpers.go:
```diff
@@ -29,3 +32,20 @@ func HasLabel(label string, issueLabels []Label) bool {
 }
+
+// ImageSmallerThan checks if image in url is smaller in size than limit
+func ImageSmallerThan(url string, limit int64) (bool, error) {
```
Since GitHub has a fixed image size limit the `limit` parameter is unnecessary.

cjwagner commented on prow/plugins/dog/dog_test.go:
```diff
@@ -96,7 +96,7 @@ func TestFormat(t *testing.T) {
 func TestHttpResponse(t *testing.T) {
 	// setup a stock valid request
-	url := "http://localhost/dog.jpg"
+	url := "https://random.dog/2e0de7b5-cc16-4112-8c5e-2c18708cf800.jpg"
```
I'm not sure if this is allowable, it is not hermetic. @BenTheElder

BenTheElder replied:
this will break if the site is down or similar, we can fake out the request locally which will be less flaky..

matthyx replied:
Ok, I will use httptest instead. That was actually my question in the comment...
I plan also to make cat_test and dog_test more similar.
(not sure I can work on that before next week)

BenTheElder replied:
Sorry, I receive a *lot* of email so sometimes it's difficult to keep up with and respond to all GitHub comments (which are a lot of that email themselves)...

Using real images for the fake request would be somewhat reasonable, but only if they were checked in. Making real HTTP requests in tests is not ideal.

matthyx commented:
No worries :-) I will see what I can do with httptests and let you know.
Thanks for your help!

On Wed, May 16, 2018, 23:53 Benjamin Elder <notifications@github.com> wrote:

> *@BenTheElder* commented on this pull request.
> ------------------------------
>
> In prow/plugins/dog/dog_test.go
> <https://github.com/kubernetes/test-infra/pull/8034#discussion_r188784248>
> :
>
> > @@ -96,7 +96,7 @@ func TestFormat(t *testing.T) {
>  // Medium integration test (depends on ability to open a TCP port)
>  func TestHttpResponse(t *testing.T) {
>  	// setup a stock valid request
> -	url := "http://localhost/dog.jpg"
> +	url := "https://random.dog/2e0de7b5-cc16-4112-8c5e-2c18708cf800.jpg"
>
> Sorry, I receive a *lot* of email so sometimes it's difficult to keep up
> with and respond to all GitHub comments (which are a lot of that email
> themselves)...
>
> Using real images for the fake request would be somewhat reasonable, but
> only if they were checked in. Making real HTTP requests in tests is not
> ideal.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/test-infra/pull/8034#discussion_r188784248>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ATuakQBwkav8Vb2ESI2pVdEgbk8OqeIdks5tzJ_ngaJpZM4T-91w>
> .
>

matthyx commented:
@cjwagner I have removed the limit parameter and used a local httptest server to get the Content-Length headers for dogs.
If that's fine like that, I will modify the cats too.

cjwagner commented on prow/plugins/dog/dog_test.go:
```diff
@@ -151,6 +171,23 @@ func TestHttpResponse(t *testing.T) {
+		}
+	}))
+	ts2.Listener.Close()
+	ts2.Listener = listener
```
I don't think there is a need to create a separate listener to replace the one provided by the `httptest` server.
It looks like you are doing this so that you can determine the port that the listener is listening to, but you can instead just use the `httptest` server's `URL` field.

matthyx replied:
Well, there is a chicken/egg problem here... as I need the port when declaring the handler function, and ts2.URL is only available after I have already created/started the server by specifying the handler.

cjwagner replied:
The port is used by the handler func for `ts` not `ts2`. `ts2`'s handler only captures `contentLength` which doesn't depend on the port. Couldn't `ts2` just be defined before everything else and it's URL used directly instead of replacing the listener?

matthyx replied:
Oh... yes! This is a remain from when I was planning to use only one server.
You are 100% right, I will correct that and use the same strategy for cats if that's ok for you?

cjwagner replied:
Yes, that would be great!

cjwagner commented on prow/plugins/cat/cat_test.go:
```diff
@@ -175,16 +177,81 @@ func TestFormat(t *testing.T) {
+	}))
+	defer ts2.Close()
+	// get the port to construct image urls
+	port := ts2.Listener.Addr().(*net.TCPAddr).Port
```
You don't need to extract the port like this. `ts2.URL` provides the URL of the server in the form `http://<ip>:<port>` so below you can specify `img := ts2.URL + "/cat.jpg"`...

matthyx commented:
Thanks @cjwagner, I have pushed this small change to both tests.
